bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,766 @@
+                    };
+                }
+            }
+
+            if !params.arbitary_spks.is_empty() {
+                let mut arbitary_spks = params
+                    .arbitary_spks
+                    .iter()
+                    .enumerate()
+                    .map(|(i, spk)| (i, spk.clone()));
+                match self.populate_with_spks::<K, _, _>(
+                    &mut update,
+                    &mut arbitary_spks,
+                    usize::MAX,
+                    params.batch_size,
+                ) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_) => {}
+                }
+            }
+
+            if !params.txids.is_empty() {
+                match self.populate_with_txids(&mut update, params.txids.iter().cloned()) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_) => {}
+                }
+            }
+
+            if !params.full_txs.is_empty() {
+                match self.populate_with_txs(&mut update, params.full_txs.iter()) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_) => {}
+                }
+            }
+
+            if !params.outpoints.is_empty() {
+                match self.populate_with_outpoints(&mut update, params.outpoints.iter().cloned()) {
+                    Err(InternalError::Reorg) => continue,
+                    Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                    Ok(_) => {}
+                }
+            }
+
+            match self.reorg_check(&mut update) {
+                Err(InternalError::Reorg) => continue,
+                Err(InternalError::ElectrumError(e)) => return Err(e.into()),
+                Ok(_) => break update,
+            }
+        };
+
+        let last_active_index = request_spks
+            .into_keys()
+            .filter_map(|k| {
+                scanned_spks
+                    .range((k.clone(), u32::MIN)..=(k.clone(), u32::MAX))
+                    .rev()
+                    .find(|(_, (_, active))| *active)
+                    .map(|((_, i), _)| (k, *i))
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        Ok(ScanUpdate {
+            update,
+            last_active_indices: last_active_index,
+        })
+    }
+}
+
+/// Parameters for [`ElectrumClient::scan`].
+pub struct ScanParams<K, S: IntoIterator<Item = (u32, Script)>> {
+    /// Indexed script pubkeys of each keychain to scan transaction histories for.
+    pub keychain_spks: BTreeMap<K, S>,
+    /// Arbitary script pubkeys to scan transaction histories for, which are not associated to a
+    /// keychain.
+    pub arbitary_spks: Vec<Script>,
```

Comment:
We need this to be an iterator too so we know what progress is being made. It should be called `misc_spks` rather than arbitrary (they are not arbitrary in most senses). 

I would still prefer not to have this and force the caller to put this into a keychain. You can have a convenience method which creates a `ScanParams` from an iterator of `Script`.

Reply:
Done in c1a0443999af9842a16cf9f92c26250e79189f0d